presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java

```diff
@@ -17,210 +31,667 @@
+    private final int[] order;
+
+    public TDigest(double compression)
+    {
+        // ensure compression >= 10
+        // default size = 2 * ceil(compression)
+        // default bufferSize = 5 * size
+        // scale = max(2, bufferSize / size - 1)
+        // compression, publicCompression = sqrt(scale-1)*compression, compression
+        // ensure size > 2 * compression + weightLimitFudge
+        // ensure bufferSize > 2*size
+
+        if (compression < 10) {
+            compression = 10;
+        }
+
+        double sizeFudge = 10;
+        if (compression < 30) {
+            sizeFudge += 20;
+        }
+
+        // publicCompression is how many centroids the user asked for
+        // compression is how many we actually keep
+        this.publicCompression = compression;
+
+        int size = (int) Math.ceil(this.publicCompression + sizeFudge);
+        int bufferSize = 5 * size;
+
+        // scale is the ratio of extra buffer to the final size
+        // we have to account for the fact that we copy all live centroids into the incoming space
+        double scale = Math.max(1, bufferSize / size - 1);
+
+        this.compression = Math.sqrt(scale) * publicCompression;
+
+        size = (int) Math.ceil(this.compression + sizeFudge);
+
+        weight = new double[size];
+        mean = new double[size];
+
+        tempWeight = new double[bufferSize];
+        tempMean = new double[bufferSize];
+        order = new int[bufferSize];
+
+        activeCentroids = 0;
     }
 
-    /**
-     * Creates a TDigest of whichever type is the currently recommended type.  MergingDigest is generally the best
-     * known implementation right now.
-     *
-     * @param compression The compression parameter.  100 is a common value for normal uses.  1000 is extremely large.
-     *                    The number of centroids retained will be a smallish (usually less than 10) multiple of this number.
-     * @return the TDigest
-     */
-    @SuppressWarnings({"unused", "WeakerAccess", "SameParameterValue"})
-    public static TDigest createDigest(double compression) {
-        return createMergingDigest(compression);
+    public TDigest(Slice slice)
+    {
+        SliceInput sliceInput = new BasicSliceInput(slice);
+        byte format = sliceInput.readByte();
+        checkArgument(format == 0, "Invalid format");
+        byte type = sliceInput.readByte();
+        checkArgument(type == 0, "Invalid type; expecting '0' (type double)");
+        this.min = sliceInput.readDouble();
+        this.max = sliceInput.readDouble();
+        this.publicCompression = sliceInput.readDouble();
+        this.totalWeight = sliceInput.readDouble();
+        this.activeCentroids = sliceInput.readInt();
+
+        this.queryOnly = true;
+        double sizeFudge = 10;
+        if (this.publicCompression < 30) {
+            sizeFudge += 20;
+        }
+
+        int size = (int) Math.ceil(this.publicCompression + sizeFudge);
```

Comment:
@rongrong Ideally, we would avoid repeating code. However, our main constructor takes in a compression parameter (double). We could call that constructor with `this()`, but we still do not know what the compression factor from the serialized t-digest is. Also, `this()` must be the first thing you call in your function for Java to allow it. Since we first have to deserialize the Slice to figure out the compression factor, we can't use `this()`. I tried repeating the least amount of code possible. Alternatively, if you prefer, I could make a helper function that has all the repeated code, and just call that helper from both functions.

Reply:
In that case I think a helper function is better (give it a more meaningful name like `initializeXXX` rather than call it helper though. Another way is to have a static create function to do the deserialization before instantiating the object.